posts/2021-07-28-GATs-stabilization-push.md

```diff
@@ -9,56 +9,87 @@ team: the Traits Working Group <https://www.rust-lang.org/governance/teams/compi
 
 ## So what are GATs?
 
 **Note: this will only be a brief overview. The [RFC] contains many more details.**
 
-GATs (generic associated types) were originally proposed in [RFC 1598][RFC]. Put simply, they allow you to define type, lifetime, or const generics on associated types. If you're familiar with languages that have "higher-kinded types", then you could calls GATs *type constructors on traits*.
+GATs (generic associated types) were originally proposed in [RFC 1598][RFC]. Put simply, they allow you to define type, lifetime, or const generics on associated types. If you're familiar with languages that have "higher-kinded types", then you could call GATs *type constructors on traits*.
 
 To jump into an example of GATs, here is a popular use case: a `LendingIterator` (formerly known as a `StreamingIterator`):
 
 ```rust
 trait LendingIterator {
-    type Item<'a>;
+    type Item<'a> where Self: 'a;
     fn next<'a>(&'a mut self) -> Option<Self::Item<'a>>;
 }
 ```
 
-With this, you could create a line iterator over a `String` without cloning each line...
+Let's go through one implementation of this, a hypothetical `<[T]>::windows_mut`, which allows for iterating through overlapping mutable windows on a slice. If you were to try to implement this with `Iterator` today like
 ```rust
-struct LineIterator { ... }
-impl LendingIterator for LineIterator {
-    type Item<'a> = &'a str;
-    ...
+struct WindowsMut<'t, T: 't> {
+    slice: &'t mut [T],
+    start: usize,
+    window_size: usize,
+}
+impl<'t, T: 't> Iterator for WindowsMut<'t, T> {
+    type Item = &'t mut [T];
+    fn next<'a>(&'a mut self) -> Option<Self::Item> {
+        let retval = self.slice[self.start..].get_mut(..self.window_size)?;
+        self.start += 1;
+        Some(retval)
+    }
 }
-fn lines(text: String) -> LineIterator { ... }
-// later...
-let all_lines = lines(String::from("A\nmulti\nline\nString!\n"));
-let first_line: &str = all_lines.next();
-assert_eq!(first_line, "A");
 ```
+then you would get an error.
+```rust
+error[E0495]: cannot infer an appropriate lifetime for lifetime parameter in function call due to conflicting requirements
+  --> src/lib.rs:9:22
+   |
+9  |         let retval = self.slice[self.start..].get_mut(..self.window_size)?;
+   |                      ^^^^^^^^^^^^^^^^^^^^^^^^
+   |
+note: first, the lifetime cannot outlive the lifetime `'a` as defined on the method body at 8:13...
+  --> src/lib.rs:8:13
+   |
+8  |     fn next<'a>(&'a mut self) -> Option<Self::Item> {
+   |             ^^
+note: ...so that reference does not outlive borrowed content
+  --> src/lib.rs:9:22
+   |
+9  |         let retval = self.slice[self.start..].get_mut(..self.window_size)?;
+   |                      ^^^^^^^^^^
+note: but, the lifetime must be valid for the lifetime `'t` as defined on the impl at 6:6...
+  --> src/lib.rs:6:6
+   |
+6  | impl<'t, T: 't> Iterator for WindowsMut<'t, T> {
+   |      ^^
+```
+
```

Comment:
Can you add one sentence summarizing the error? 

"The errors above indicate that the lifetime of `&mut self` (known as `'a`) and the lifetime of `Self::Item` (known as `'t`) are unrelated and the compiler cannot prove that `'a` lives as long as `'t` does. Without this proof, the item returned from the iterator might be allowed to live longer than the slice we're iterating over!"

Reply:
hmm, I would have stated it in terms of `impl Iterator for WindowsMut` would end up creating overlapping mutable references, since `Item`'s lifetime doesn't prevent calling `next` again while holding the previously returned reference. `LendingIterator` does prevent calling `next` until the previously returned `Item` is dropped.